srml/balances/src/lib.rs

```diff
@@ -1150,12 +1153,33 @@ where
 #[derive(Encode, Decode, Clone, Eq, PartialEq)]
 pub struct TakeFees<T: Trait<I>, I: Instance = DefaultInstance>(#[codec(compact)] T::Balance);
 
-#[cfg(feature = "std")]
 impl<T: Trait<I>, I: Instance> TakeFees<T, I> {
 	/// utility constructor. Used only in client/factory code.
+	#[cfg(feature = "std")]
 	pub fn from(fee: T::Balance) -> Self {
 		Self(fee)
 	}
+
+	/// Compute the final fee value for a particular transaction.
+	///
+	/// The final fee is composed of:
+	///   - _length-fee_: This is the amount paid merely to pay for size of the transaction.
+	///   - _weight-fee_: This amount is computer based on the weight of the transaction. Unlike
+	///      size-fee, this is not input dependent and reflects the _complexity_ of the execution
+	///      and the time it consumes.
+	///   - (optional) _tip_: if included in the transaction, it will be added on top. Only signed
+	///      transactions can have a tip.
+	fn compute_fee(len: usize, weight: Weight, tip: T::Balance) -> T::Balance {
+		// length fee
+		let len = T::Balance::from(len as u32);
+		let len_fee = T::TransactionBaseFee::get()
```

Comment:
should be mutated based on `DispatchInfo`. i.e. introduce `DispatchInfo::get_base_fee(len: usize)`. for `Free` dispatch classes, this could then be reduced to zero.

Reply:
I thought of such a design a while ago but the problem for that would be that all of the parameters need to be passed in, since the `DispatchInfo` does not have access to them. It will be: `DispatchInfo::get_base_fee<B>(len: usize, base_fee: B, byte_fee: B)`

Reply:
(unless if we want to move fee-related associated constants stuff into to `core`, which I don't see good)

Reply:
I think the best is to set the design as follows: 

Signed Extension only sees `DispatchClass` and won't interact with `SimpleDispatchInfo` directly (since it actually can't see it). If you want _truly zero-cost transactions_: This should be reflected in a new `DispatchClass::UserFree , DispatchClass::OperationalFree`. A bit verbose but keeps things where they belong.

btw. I am not sure if we want such transactions actually. You can have a `zero-weight-fee`, the decision to bring back the traditional `base + byte` is what is preventing us from having it. It might still be sensible to say: _you might get away with the weight fee, but you always pay byte fee (it should be smaller anyhow)_.